dpnp/dpnp_iface_manipulation.py

```diff
@@ -245,7 +264,122 @@ def _calc_parameters(a, axis):
     n = a.shape[axis]
     a_shape = list(a.shape)
 
-    return a, a_ndim, order, axis, slobj, n, a_shape
+    if dpnp.is_supported_array_type(obj) and dpnp.is_supported_array_type(
+        values
+    ):
+        usm_type, exec_q = get_usm_allocations([a, obj, values])
+    elif dpnp.is_supported_array_type(values):
+        usm_type, exec_q = get_usm_allocations([a, values])
+    elif dpnp.is_supported_array_type(obj):
+        usm_type, exec_q = get_usm_allocations([a, obj])
+    else:
+        usm_type, exec_q = a.usm_type, a.sycl_queue
+
+    return Parameters(
+        a, a_ndim, order, axis, slobj, n, a_shape, exec_q, usm_type
+    )
+
+
+def _insert_array_indices(parameters, indices, values, obj):
+    """
+    Utility function for ``dpnp.insert`` when indices is an array with
+    multiple elements.
+
+    """
+
+    a, a_ndim, order, axis, slobj, n, a_shape, exec_q, usm_type = parameters
+
+    is_array = isinstance(obj, (dpnp_array, numpy.ndarray, dpt.usm_ndarray))
+    if indices.size == 0 and not is_array:
+        # Can safely cast the empty list to intp
+        indices = indices.astype(dpnp.intp)
+
+    indices[indices < 0] += n
+
+    numnew = len(indices)
+    ind_sort = indices.argsort(kind="stable")
+    indices[ind_sort] += dpnp.arange(
+        numnew, dtype=indices.dtype, sycl_queue=exec_q, usm_type=usm_type
+    )
+
+    a_shape[axis] += numnew
+    old_mask = dpnp.ones(
+        a_shape[axis], dtype=dpnp.bool, sycl_queue=exec_q, usm_type=usm_type
+    )
+    old_mask[indices] = False
+
+    new = dpnp.empty(
+        a_shape,
+        dtype=a.dtype,
+        order=order,
+        sycl_queue=exec_q,
+        usm_type=usm_type,
+    )
```

Comment:
Can it be simplified?
```suggestion
    new = dpnp.empty_like(a, order=order)
```

Reply:
Other inputs (`obj` and `values`) may also play a role in determining `sycl_queue` and `usm_type` and should be passed to `empty` or `empty_like`. So, it seems there is no simplification using one over the other one.